ChainSafe/gossamer
# chore(lib/runtime): refactor version struct and codec

## Changes

- Version code
  - Single `Version` struct
  - Encoding the `Version` always encode all the last fields
  - Decoding the version data to `Version` sets possible absent fields as zero if not in the data (`transaction_version`)
  - Remove `LegacyVersionData` struct, `Version` interface, getter methods, cloned structs for codec
  - Move decoding logic from `lib/runtime/wasmer/exports.go` to `lib/runtime/version.go` in `DecodeVersion` function
  - Use struct value for `Version` instead of its pointer (no reason to use a pointer)
- Test code
  - Refactor multiple `TestInstance_Version*` in one test with test cases
  - Add runtime API items as expected fields in tests
  - Remove version runtime mockery mock
  - Simplify testing in other packages since we no longer need to use a test-only exported constructor `NewVersionData`

## Tests

## Issues

I had to change a whole bunch of version related code to persist the state version, and I thought I would simplify it since it was rather confusing as a reader.

## Primary Reviewer

@timwu20
